### ISSUE TITLE
Switch kernel build to `linuxkit pkg`

### ISSUE THREAD

**Issue by ijc** [linuxkit/linuxkit#2583]
Once #2578 is put to bed I'd like to convert the kernel build over.

Looking at `kernel/Makefile` I can see 4 possible issues which may need enhancements to `linuxkit pkg`. 

1. Templating to handle multiple kernel versions. I think this will be fine, it'll be multiple invocations of `linuxkit pkg` against either a statically or automatically generated `build-foo.yml`.
2. The use of `--build-arg`. Should be simple enough to add a similar argument to `linuxkit pkg build` (and `push`) which is propagated to the build.
3. Invokes `docker push` and `push-manifest` more than once. This is because it builds and pushes `linuxkit/kernel:X.Y.Z$(EXTRA)-$(TAG)` by default (`X.Y.Z` is the kernel version, `$(EXTRA)` is an optional per image suffix like `-dbg`, `$(TAG)` is the usual tree-sh and `-dirty` suffix) but also tags and pushes `linuxkit/kernel:X.Y.Z$(EXTRA)` (no hash). This differs from the usual pattern of `linuxkit/pkg:$(TAG)`. For this we could add a new field to the `build.yml` called `tag: foo` where `foo` here would be `X.Y.Z$(EXTRA)` (per-kernel). If this field is set then the build/push will be of both `foo-$(TAG)` and `foo` (but not the default `$(TAG)`. Other names such as `alias` or `prefix` might be tempting but don't seem to match the semantics (in particular because the "default" name is not used so it is neither an alias nor is it strictly a prefix).
4. Secondary builds (`perf`, `zfs`) which use the just-now-built kernel image. I think the `build-arg` stuff above handles that?

/cc @rn any of that sound insane? Any other wrinkles with the kernel build I've missed?

**Comment by rn** [comment]
- I think having `tag: foo` in the `build.yml` is fine. if present we should push to `<org>/<img>:<tag>` *and* `<org>/<img>:<tag>-<hash>`
- Specifying `--build-arg` in the `build.yml` would be good, I think. I prefer if I only have to change the kernel version in two places (currently `Makefile` and `kernel-config-X`). If this is an argument to `linuxkit` and is in `build.yml` it'll be three files.
- The secondary builds are tricky as they require the hash from the main kernel build. There is added complication as one can't do `docker build` with content trust where the `FROM` line comes from a build arg. It's a bug, so we do a trusted pull first (which weirdly works). This will be tricky/special case for `linuxkit pkg`. See big comment just above `build_perf` in `Makefile`

**Comment by ijc** [comment]
> * I think having tag: foo in the build.yml is fine. if present we should push to <org>/<img>:<tag> and <org>/<img>:<tag>-<hash>

:+1:

> * Specifying --build-arg in the build.yml would be good, I think. I prefer if I only have to change the kernel version in two places (currently Makefile and kernel-config-X). If this is an argument to linuxkit and is in build.yml it'll be three files.

Good idea, lets have the tool read an array of build args from a `yml` key (lets call it `build-args`).

That seems less open to abuse or mistakes (such as not putting the command invocation line in git and therefore missing it in the hash) than a CLI option, although maybe we will eventually find a use case for that too.

> * The secondary builds are tricky as they require the hash from the main kernel build. There is added complication as one can't do docker build with content trust where the FROM line comes from a build arg. It's a bug, so we do a trusted pull first (which weirdly works). This will be tricky/special case for linuxkit pkg. See big comment just above build_perf in Makefile

Will have to think on this one some more. Perhaps if I do so for long enough the bug will be fixed ;-) (is there an issue on moby/moby for it?),

Perhaps we can just generate `Dockerfile.perf` from a template with sed?

**Comment by rn** [comment]
- moby issue linked in the comment: https://github.com/moby/moby/issues/34199
- `sed` is certainly an option

**Comment by ijc** [comment]
It is valid to do e.g. `COPY --from=docker:17.06.0-ce` rather than having to do `FROM docker:17.06.0-ce AS docker` then `COPY --from=docker`. Unfortunately `COPY --from=${IMAGE}` doesn't work:
```
Step 5/5 : COPY --from=${IMAGE} /perf /usr/bin/perf
invalid from flag value ${IMAGE}: invalid reference format: repository name must be lowercase
Makefile:175: recipe for target 'build_perf_4.9.x' failed
```
Which is a shame. Even if it did work I expect there would be some chance it would also be incompatible with DCT.

**Comment by ijc** [comment]
Overall looks good.

The repetition of the versions you mention could be resolved by having e.g. `4.14.x/{kernel,dbg,perf}` instead of ``4.14.x{,-dbg,-perf}`. Or you could have them all in the same dir but differently named `build.yml` files so you would use `linuxklit pkg build --build-yml=build-perf.yml` to build the variants. Might also allow you to have the same hashes?

The `Extend the build-args, introduced above to use a special value` thing seems a bit too special to me. Is that just for the `FROM` used in the derivative image? Perhaps a distinct `parent-images` section in the yaml would be less magic? i.e. a map from a build arg name to an image name which then gets set much as you proposed (it's just now more distinct from a regular build-arg and also more targetted).

I wonder if adding support for multi-stage builds to `build.yml` (i.e. a `target` field) would help you here at all, allowing you to combine into a single `Dockerfile` are reducing the need for cross play between different dockerfiles?

**Comment by rn** [comment]
> The repetition of the versions you mention could be resolved by having e.g. 4.14.x/{kernel,dbg,perf} instead of ``4.14.x{,-dbg,-perf}. Or you could have them all in the same dir but differently namedbuild.yml` files so you would use `linuxklit pkg build --build-yml=build-perf.yml` to build the variants. Might also allow you to have the same hashes?

I think this is a good idea. Same dir, different `build.yml`. They will still have different hashes as the `extra-sources` will be different.

> The Extend the build-args, introduced above to use a special value thing seems a bit too special to me. Is that just for the FROM used in the derivative image? Perhaps a distinct parent-images section in the yaml would be less magic? i.e. a map from a build arg name to an image name which then gets set much as you proposed (it's just now more distinct from a regular build-arg and also more targetted).

I like this. Definitely better than the special value hack which I wasn't too happy with in the first place.

> I wonder if adding support for multi-stage builds to build.yml (i.e. a target field) would help you here at all, allowing you to combine into a single Dockerfile are reducing the need for cross play between different dockerfiles?

That might be a good idea for `perf` and `zfs` (and `bcc` once merged). And I *think* this might result in most cases for the related packages to have the same hash as well.

**Comment by ijc** [comment]
> They will still have different hashes as the `extra-sources` will be different.

I don't think that _has_ to be the case, it would just mean some of them had some redundant or unnecessary entries, I guess. The multi-stage target thing would have the same effect I guess.
